crates/shadowsocks-service/src/local/tun/tcp.rs

```diff
@@ -0,0 +1,272 @@
+    loadbalancing::PingBalancer,
+    net::AutoProxyClientStream,
+    utils::{establish_tcp_tunnel, to_ipv4_mapped},
+};
+
+struct TcpAddressTranslator {
+    connections: LruCache<SocketAddr, TcpConnection>,
+    mapping: LruCache<(SocketAddr, SocketAddr), SocketAddr>,
+}
+
+impl TcpAddressTranslator {
+    fn new() -> TcpAddressTranslator {
+        TcpAddressTranslator {
+            connections: LruCache::with_expiry_duration(Duration::from_secs(24 * 60 * 60)),
+            mapping: LruCache::with_expiry_duration(Duration::from_secs(24 * 60 * 60)),
+        }
+    }
+}
+
+pub struct TcpTun {
+    tcp_daddr: SocketAddr,
+    free_addrs: Vec<IpAddr>,
+    translator: Arc<Mutex<TcpAddressTranslator>>,
+    abortable: JoinHandle<io::Result<()>>,
+}
+
+impl Drop for TcpTun {
+    fn drop(&mut self) {
+        self.abortable.abort();
+    }
+}
+
+impl TcpTun {
+    pub async fn new(context: Arc<ServiceContext>, tun_network: IpNet, balancer: PingBalancer) -> io::Result<TcpTun> {
+        let mut hosts = tun_network.hosts();
+        let tcp_daddr = match hosts.next() {
+            Some(d) => d,
+            None => return Err(io::Error::new(ErrorKind::Other, "tun network doesn't have any hosts")),
+        };
+
+        // Take up to 10 IPs as saddr for NAT allocating
+        let free_addrs = hosts.take(10).collect::<Vec<IpAddr>>();
+        if free_addrs.is_empty() {
+            return Err(io::Error::new(
+                ErrorKind::InvalidInput,
+                "tun network doesn't have enough free addresses",
+            ));
+        }
+
+        let listener = TcpListener::bind_with_opts(&SocketAddr::new(tcp_daddr, 0), context.accept_opts()).await?;
+        let tcp_daddr = listener.local_addr()?;
+
+        debug!("tun tcp listener bind {}", tcp_daddr);
+
+        let translator = Arc::new(Mutex::new(TcpAddressTranslator::new()));
+
+        let abortable = {
+            let translator = translator.clone();
+            tokio::spawn(TcpTun::tunnel(context, listener, balancer, translator))
+        };
+
+        Ok(TcpTun {
+            tcp_daddr,
+            free_addrs,
+            translator,
+            abortable,
+        })
+    }
+
+    pub async fn handle_packet(
+        &mut self,
+        src_addr: SocketAddr,
+        dst_addr: SocketAddr,
+        tcp_header: &TcpHeader,
+    ) -> io::Result<Option<(SocketAddr, SocketAddr)>> {
+        let TcpAddressTranslator {
+            ref mut connections,
+            ref mut mapping,
+        } = *(self.translator.lock().await);
+
+        let (conn, is_reply) = if tcp_header.syn && !tcp_header.ack {
+            // 1st SYN, creating a new connection
+            // Allocate a `saddr` for it
+            let saddr = loop {
+                let addr_idx = rand::random::<usize>() % self.free_addrs.len();
+                let port = rand::random::<u16>() % (65535 - 1024) + 1024;
+
+                let addr = SocketAddr::new(self.free_addrs[addr_idx], port);
```

Comment:
Why we use random up to ten saddrs to determinate which saddr to used? I think use fixed one saddr is enough

Reply:
Well, then you can have at most `(65535 - 1024) * 10` free addresses.

Reply:
Now the `tun`'s address is included in the free addresses.